# Path of Exile game stats data eda

dates: 2022.04.20 ~ 

This file contains an analysis processing of the game 'Path of Exile' data.

###### Origins of datas from here: 
poe_stats : https://www.kaggle.com/datasets/gagazet/path-of-exile-league-statistic

Data contains stats of 59000 players, from 4th August of 2017 and before now.




# Dataset exploring

## data and libraries import

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

poe_stats = pd.read_csv('poe_stats.csv')

In [2]:
poe_stats

rank   dead  online                    name  level        class  \
0          1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1          1  False   False        RaizNeverFirstQT    100  Necromancer   
2          1  False   False  GucciStreamerAdvantage    100  Necromancer   
3          1  False   False             ChiroxPrime    100       Slayer   
4          2  False   False     Cool_NecroIsFineNow    100      Deadeye   
...      ...    ...     ...                     ...    ...          ...   
59771  14999   True   False       ПроклятьеРекласта     89  Necromancer   
59772  15000  False   False       IshibashiSummoner     94  Necromancer   
59773  15000  False   False                 BLively     73       Slayer   
59774  15000  False   False                vawddvaw     89    Gladiator   
59775  15000  False   False                 Reselin     53  Necromancer   

                                                      id  experience  \
0      3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1      8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2      c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3      c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4      24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
...                                                  ...         ...   
59771  d33b4f6e08c10e365765f9a36a8f36d561fd1d86f10ee1...  1786695653   
59772  5764cfa387e0a87a4bebc1a3c5017e92de8bbb06445c47...  2720770853   
59773  9ac75ab75a47cee8a9dfb0a31912df8909720a8b20f9f1...   495034029   
59774  cf02dfc0c90b2df9c7ac76bbedd91e93c2a8a2ca629c15...  1786638972   
59775  f7ffda5ca2490546344d32930693f8299930ce9451df4a...    77909816   

            account  challenges        twitch              ladder  
0            TheTzn          40        thetzn           Harbinger  
1            RaizQT          22        raizqt    SSF Harbinger HC  
2       GucciPradas          35   guccipradas  Hardcore Harbinger  
3          Chiroxun          26           NaN       SSF Harbinger  
4          cooltail          22           NaN           Harbinger  
...             ...         ...           ...                 ...  
59771  Romanchik002          18  romanchik002  Hardcore Harbinger  
59772  Ishibashisan          33           NaN           Harbinger  
59773        ruby01          19           NaN       SSF Harbinger  
59774        Lecarn          24        lecarn  Hardcore Harbinger  
59775      woworpoe          19           NaN    SSF Harbinger HC  

[59776 rows x 12 columns]

In [3]:
df = poe_stats
df.sample(5)

rank   dead  online             name  level       class  \
468     118  False   False      KAN_lowtier    100  Inquisitor   
12406  3109   True   False        Shockvoid     82   Trickster   
10811  2709  False   False       ANA_A_SRay     89   Trickster   
10771  2699  False   False             Xixa     97      Raider   
13265  3324  False   False  Kolbenkoenig_BG     97   Berserker   

                                                      id  experience  \
468    370f2104ab7f712d3a65249948366ba4b9206226e3491c...  4250334444   
12406  9141654c01e11a6129f49e345965006ac44aeedca62c37...  1055801342   
10811  fac1b9944472d529d0f6a60b49c41b10c8aa42a16d5d14...  1825004357   
10771  d9869bad1ad4616cb6c8562f1817cc1e89bbcb5cc834bd...  3386297529   
13265  70b8876dfca3d67959651eb14535236602ab1c371d2db6...  3365382202   

         account  challenges   twitch              ladder  
468     boomy212          25      NaN  Hardcore Harbinger  
12406    nazgarn          13  nazgarn    SSF Harbinger HC  
10811  cjhhelpme          33      NaN       SSF Harbinger  
10771     Xiphon          26      NaN           Harbinger  
13265      pfloo          38      NaN           Harbinger

In [4]:
df.shape

(59776, 12)

## data describes

In [5]:
df.describe()

rank         level    experience    challenges
count  59776.000000  59776.000000  5.977600e+04  59776.000000
mean    7495.369379     85.422578  1.825786e+09     25.759569
std     4330.179898     11.288017  1.094748e+09      8.440888
min        1.000000     53.000000  7.790982e+07      4.000000
25%     3745.000000     78.000000  7.774697e+08     19.000000
50%     7492.000000     90.000000  1.934010e+09     25.000000
75%    11247.000000     95.000000  2.876117e+09     33.000000
max    15000.000000    100.000000  4.250334e+09     40.000000

In [6]:
df.describe(include='object')

name      class                                                 id  \
count    59776      59776                                              59776   
unique   58286         26                                              59776   
top     _Flic_  Berserker  3dcddd59f5088893f734f39686350990dae168cc4f4b00...   
freq        14       8711                                                  1   

        account            twitch         ladder  
count     59776             16300          59776  
unique    37123              9613              4  
top     chuz101  thespunkylobster  SSF Harbinger  
freq         30                18          14981

In [7]:
df.columns

Index(['rank', 'dead', 'online', 'name', 'level', 'class', 'id', 'experience',
       'account', 'challenges', 'twitch', 'ladder'],
      dtype='object')

In [8]:
df.nunique()

rank          15000
dead              2
online            2
name          58286
level            48
class            26
id            59776
experience    54124
account       37123
challenges       37
twitch         9613
ladder            4
dtype: int64

In [9]:
df.isnull().sum()

rank              0
dead              0
online            0
name              0
level             0
class             0
id                0
experience        0
account           0
challenges        0
twitch        43476
ladder            0
dtype: int64

twitch 아이디가 없는 사람도, 혹은 한 twitch 아이디에 여러 계정을 가지고 있는 사람도 있을 것이다. 

In [10]:
df['twitch'].describe()

count                16300
unique                9613
top       thespunkylobster
freq                    18
Name: twitch, dtype: object

In [11]:
df['twitch'].groupby(df['twitch']).count().sort_values()

twitch
001100010010         1
newromancers03       1
newfy1991            1
newbieroadkill       1
newancestry          1
                    ..
yoshixt             15
zen1703             15
iispasov            16
seyser_koze         16
thespunkylobster    18
Name: twitch, Length: 9613, dtype: int64

### describe results

데이터 총 개수 59776 개

컬럼명: ['rank', 'dead', 'online', 'name', 'level', 'class', 'id', 'experience', 'account', 'challenges', 'twitch', 'ladder']

랭크는 총 1만 5천개, 죽음과 온라인 여부는 True, False로 이루어져있다.

id가 오직 unique 하며, 한 아이디당 여러 account를 가질 수 있는 것으로 보인다. 그에 따라 일부 name은 겹치는 것을 보인다.

level 과 class는 게임 내 레벨과 직업으로 이루어져 있다.

experience 는 게임 내 경험치를 의미하는 것일 것이며,

twitch는 트위치 계정이 적혀있는 컬럼으로 보인다. 일부 살펴본 결과 한 트위치 아이디로 여러 id를 운영하며 스트리밍 하는 것으로 보인다. 최대 18 아이디를 운영하는 twitch id도 있다.

challenges 의 경우 poe 게임이 던전형 rpg임을 감안하여 짐작해보건대 해당 랭크를 따기까지 해당 모드를 몇 번 플레이 했는지를 나타내는 것으로 예상된다.

##### ladder의 경우 

고유 값이 4개로, 원본 데이터 문서를 찾아본 결과 

```
One file with 12 data sections. One league - Harbinger, but 4 different types of divisions:
Harbinger
Hardcore Harbinger
SSF Harbinger
SSF Harbinger HC
Each division has own ladder with leaders
```
라고 설정되어있다. 

https://pathofexile.fandom.com/wiki/Harbinger_league#Challenges_and_Rewards 와 
https://pathofexile.fandom.com/wiki/Solo_Self-Found 를 참조한 결과 
> Harbinger라는 이름의 시즌 리그의 일종으로 보이며 HC는 Hardcore의 약어로 보인다. Hardcore Harbinger에서 죽은 경우 일반 리그로 옮겨진다는 것과, SSF 가 Solo self-found 로 솔로플레이 모드(파티 거부)의 일종인 점 등을 미루어 보아 난이도는 Harbinger < Hardcore Harbinger < SSF Harbinger < SSF Harbinger HC 가 아닐까 예상된다. 물론 분석에서는 파티와 솔로 플레이 모드는 구분을 해야하며, 난이도 별로도 나눌 수 있을 것이다. 클래스 별 비율 등도 각 게임 모드마다 차이가 날 것이다.

> 또한, 해당 게임이 서버가 세계 섹터 별로 나뉘어있는지는 몰라도 세계적이든 한 지역이든 이만큼의 유저들은 랭커라고 불리울 수 있는 숫자 안에 들은 것이므로 이 분석 자체가 랭커들에 대한 분석이 될 것임을 생각해 볼 수 있다.

> 추가로, SSF가 아닌 일반 모드는 함께 플레이 하면서 랭크를 올릴 수 있기 때문에 다양한 전투 보조 직업이 랭크에 올라갈 수 있으나 SSF 모드의 경우 혼자 딜하고 보조하고 '북치고 장구치고'를 전부 해야하기 때문에 일반적인 보조 직업으로는 상위 랭크를 달성하기 어려울 것으로 예상된다.  

### EDA planning

이에 따라 현 데이터 셋은 이렇게 나누어 볼 수 있을 것이다.

1. rank (rank bins) 
2. level or exp bins
3. class
4. ladder

또한 위의 내용을 laddar 로 나누어 다시 나누어 볼 수 있을 것이다.

# meaning of 'rank' is...

현재 데이터에는 총 15000개의 랭크 고유 개수가 있는데, 데이터는 약 6만개에 이른다.

In [12]:
df['rank'].value_counts().nunique()

3

각 랭크는 최대 3명으로 되어있다. 이것만 보아도 게임 모드에 따라 랭크가 구분되어 있음을 미루어 짐작해 볼 수 있다. 

확인해보자.

In [13]:
# 각 모드의 id, rank 고유 개수

modes = ['Harbinger', 'Hardcore Harbinger', 'SSF Harbinger', 'SSF Harbinger HC']

for i in modes:
    print("ladder: ", i)
    print("values of unique id", df[df['ladder'] == i]['id'].nunique())
    print("values of unique rank", df[df['ladder'] == i]['rank'].nunique())
    print('-------------------------------')

ladder:  Harbinger
values of unique id 14918
values of unique rank 14918
-------------------------------
ladder:  Hardcore Harbinger
values of unique id 14905
values of unique rank 14905
-------------------------------
ladder:  SSF Harbinger
values of unique id 14981
values of unique rank 14981
-------------------------------
ladder:  SSF Harbinger HC
values of unique id 14972
values of unique rank 14972
-------------------------------


In [14]:
pd.pivot_table(data = df, index = ['ladder'], aggfunc = 'count')

account  challenges  class   dead  experience     id  \
ladder                                                                     
Harbinger             14918       14918  14918  14918       14918  14918   
Hardcore Harbinger    14905       14905  14905  14905       14905  14905   
SSF Harbinger         14981       14981  14981  14981       14981  14981   
SSF Harbinger HC      14972       14972  14972  14972       14972  14972   

                    level   name  online   rank  twitch  
ladder                                                   
Harbinger           14918  14918   14918  14918    4464  
Hardcore Harbinger  14905  14905   14905  14905    5452  
SSF Harbinger       14981  14981   14981  14981    2661  
SSF Harbinger HC    14972  14972   14972  14972    3723

아이디의 개수와 랭크의 개수가 일치하는 것으로 나왔다. 게임 모드에 따라 랭크가 구분되어 있음이 맞다.

개인적인 생각으로, 이 데이터는 rank별로 나열될 것이 아니라, ladder 별로 rank가 나열되어야 하는 것이 맞다고 생각이 된다.

여기서 한가지 더 해보자면 twitch 아이디가 있는 것을 따져볼 수도 있을 것이다.

## with ladder = df_byladder

In [15]:
df_byladder = df[['ladder', 'rank', 'dead', 'online', 'name', 'level', 'class', 'id', 'experience', 'account', 'challenges', 'twitch']].sort_values(by = ['ladder', 'rank'])
df_byladder.head(5)

ladder  rank   dead  online                  name  level        class  \
0   Harbinger     1  False   False    Tzn_NecroIsFineNow    100  Necromancer   
4   Harbinger     2  False   False   Cool_NecroIsFineNow    100      Deadeye   
8   Harbinger     3  False   False  Rami_The_Negromaniac    100  Necromancer   
13  Harbinger     4  False   False  Guraj_NecroIsFineNow    100       Raider   
18  Harbinger     5  False   False                 Ryphe    100    Occultist   

                                                   id  experience   account  \
0   3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444    TheTzn   
4   24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444  cooltail   
8   00c0030fc548cc4c0d804564f8fb356610044b7e6d484f...  4250334444  rami1337   
13  34bedf0408f53151dcc31eb43c3bd10a9fb0d7ba32b95b...  4250334444   Guraj93   
18  b6387a6be18939980798db2ec8e8a2924027fa803e063a...  4250334444     Ryphe   

    challenges      twitch  
0           40      thetzn  
4           22         NaN  
8           40    rami1337  
13          24  guraj_hots  
18          40  rypherious

# what 'class'es in there?

In [16]:
df_byladder['class'].unique()

array(['Necromancer', 'Deadeye', 'Raider', 'Occultist', 'Pathfinder',
       'Guardian', 'Champion', 'Gladiator', 'Inquisitor', 'Elementalist',
       'Slayer', 'Berserker', 'Juggernaut', 'Ascendant', 'Trickster',
       'Hierophant', 'Chieftain', 'Saboteur', 'Assassin', 'Ranger',
       'Marauder', 'Scion', 'Witch', 'Duelist', 'Shadow', 'Templar'],
      dtype=object)

In [17]:
df_byladder.groupby(by = ['class']).mean()

rank      dead    online      level    experience  \
class                                                                     
Ascendant      7540.713639  0.367608  0.013319  85.323122  1.785989e+09   
Assassin       8846.387334  0.331370  0.013255  80.048601  1.297246e+09   
Berserker      7601.758122  0.300884  0.013431  86.998967  1.957216e+09   
Champion       7582.100329  0.418586  0.009046  85.018914  1.690701e+09   
Chieftain      7527.660094  0.428638  0.010329  84.022535  1.633104e+09   
Deadeye        8560.173853  0.374131  0.008345  81.350487  1.411977e+09   
Duelist       12337.400000  0.753846  0.015385  61.876923  2.532442e+08   
Elementalist   8517.593514  0.333333  0.012066  80.279789  1.321416e+09   
Gladiator      7086.548367  0.427959  0.010204  84.173673  1.653267e+09   
Guardian       6277.722175  0.380629  0.015293  91.186916  2.485099e+09   
Hierophant     8034.089855  0.437681  0.003478  80.514783  1.278125e+09   
Inquisitor     7826.754940  0.385137  0.007962  83.169272  1.577004e+09   
Juggernaut     7345.391626  0.433990  0.008374  83.238424  1.552577e+09   
Marauder      10884.610526  0.631579  0.010526  64.778947  3.539756e+08   
Necromancer    7162.736329  0.383213  0.015402  84.712449  1.739417e+09   
Occultist      7617.948052  0.424513  0.004058  82.319805  1.508447e+09   
Pathfinder     6856.042935  0.087992  0.040576  94.060392  2.924074e+09   
Raider         7667.577864  0.236117  0.024660  87.094951  2.023136e+09   
Ranger        11337.166667  0.458333  0.000000  66.979167  5.301023e+08   
Saboteur       7714.562842  0.377049  0.004684  82.901639  1.545837e+09   
Scion         11941.596154  0.615385  0.000000  63.557692  3.042050e+08   
Shadow        12328.434783  0.608696  0.000000  62.869565  2.822425e+08   
Slayer         7366.899909  0.332726  0.014221  87.201641  1.983167e+09   
Templar       12541.802817  0.732394  0.000000  60.366197  1.928275e+08   
Trickster      7378.675278  0.419228  0.007521  82.643558  1.464714e+09   
Witch         12136.067227  0.655462  0.000000  62.781513  2.855004e+08   

              challenges  
class                     
Ascendant      26.141449  
Assassin       23.066274  
Berserker      26.481690  
Champion       24.095395  
Chieftain      24.358685  
Deadeye        23.168289  
Duelist        15.676923  
Elementalist   22.068627  
Gladiator      23.921633  
Guardian       30.151232  
Hierophant     21.726377  
Inquisitor     24.452374  
Juggernaut     23.739901  
Marauder       16.021053  
Necromancer    24.990957  
Occultist      24.951299  
Pathfinder     32.499646  
Raider         27.277282  
Ranger         17.645833  
Saboteur       25.249024  
Scion          14.961538  
Shadow         14.847826  
Slayer         27.014950  
Templar        13.140845  
Trickster      23.881949  
Witch          14.697479

## poe의 class 분류

해당 데이터셋에서, poe에는 ['Necromancer', 'Deadeye', 'Raider', 'Occultist', 'Pathfinder',
       'Guardian', 'Champion', 'Gladiator', 'Inquisitor', 'Elementalist',
       'Slayer', 'Berserker', 'Juggernaut', 'Ascendant', 'Trickster',
       'Hierophant', 'Chieftain', 'Saboteur', 'Assassin', 'Ranger',
       'Marauder', 'Scion', 'Witch', 'Duelist', 'Shadow', 'Templar']  총 24개의 unique 클래스가 존재하며,

POE 공식 사이트를 참고한 결과 각 직업별 클래스는 이렇게 나뉘고 있다.

    Duelist - 'Slayer', 'Gladiator', 'Champion'
    Shadow - 'Assassin', 'Saboteur', 'Trickster'
    Marauder(약탈자 라고 한다) - 'Juggernaut', 'Berserker', 'Chieftain',
    Witch - 'Necromancer', 'Elementalist', 'Occultist'
    Ranger - 'Raider', 'Deadeye', 'Pathfinder'
    Templar - 'Inquisitor', 'Hierophant', 'Guardian'
    Scion - 'Ascendant'
    
일반적으로 레인저는 원딜러, 듀얼리스트는 전사로 보이며, 위치는 마딜러, 템플러는 수도승 같은 등등 다른 게임에서 보이는 직업과 비슷한 부분이 있지만,

각 클래스가 가진 특성들도 다르기 때문에 poe에서 제공하고있는 클래스 별로도 나눌 수 있을 것이고, 일반적인 rpg 특성에 따라 이를 나누는 방법은 이런 방식들이 될 것이다.

    1. 탱커, 딜러, 힐러(보조직업)
    2. 탱커, 일반딜러, 암살딜러, 힐러(보조직업)
    2. 탱커, 근접딜러, 원거리딜러, 힐러(보조직업)
    3. 탱커, 근접딜러, 원거리물리딜러, 원거리마법딜러, 힐러(보조직업)
    4. 근거리, 원거리, 마법, 힐러(보조직업)
    5. 점사딜스킬위주, 광역딜스킬위주, 힐러(보조직업) ... 등등

이에 따라 각 클래스를 읽어보고 데이터 labeling을 진행하도록 한다.

스킬 범위나 세부 효과에 관련한 부분은 실제로 poe를 해본 사람이 분석하는 것이 더 바를 것 같다는 판단이 든다.

클래스의 호불호 혹은 성능을 떠나서, 세부적인 수치와 관련해서는 플레이 해보지 않는 이상 체감을 해볼 수 없기 때문이다.

예를 들어 템플러의 가디언은 방어 버프 혹은 적의 버프를 방해하는 스킬들이 있는데 상세히 알기는 어렵다.

---

https://hong-sam.tistory.com/27 글을 참조하여 일반적인 직업으로 나누는 것을 먼저 하기로 결정.

글에 따르면 
```
    Duelist : 빠른 근접 전투(힘, 민첩)
    Shadow : 민첩하고 지능적인 전투(민첩, 지능)
    Marauder : 전사(힘)
    Witch : 마법사(지능)
    Ranger : 궁수(민첩)
    Templar : 성기사(힘, 지능)
    Scion : 모든 스탯이 중간(힘, 민첩, 지능)
```

으로 나누어져있다. 사이온 같은 경우 성장 중 사이온을 구하는 퀘스트를 깨야 직업을 할 수 있다고 한다.

스킬을 찍는 데에 자유로운 것이 특징이나 자신의 원래 직업과 스탯에 따라 가는 것이 더 좋은 효율을 낼 수 있다는 점에서 다른 게임과 크게 다르지는 않다.

살펴본 결과 패시브 혹은 버프(디버프)스킬들도 많이 포함되어 있었다. 

자유도가 높지만 높아도 그 안에서 더 좋은 효율을 내는 일부 직업이나 스킬트리가 유행하여 해당 직업에는 한 클래스가 주를 이루기 마련인데, 그런 경향도 있는지 살펴보아야 겠다.

문제는 사이온인데, 사람마다 찍는 유형에 따라 근접일 수도 원거리일 수도, 완전히 보조 일수도, 그 중간 어딘가일 수도 있다.

여러 능력의 조합을 사용하는 것의 이점과 단점은 극명하기 때문에 숙련된 사람이 아니면 사이온을 달고 랭킹에 진입하기는 쉽지 않았을 것이라는 생각도 든다.


방식 1: 세부 클래스 전 직업(전직 전)으로 묶은 다음 그에 따라 labeling, 사이온은 사이온으로.
```
    근 - Duelist - 'Slayer', 'Gladiator', 'Champion'
    근 - Shadow - 'Assassin', 'Saboteur', 'Trickster'
    근 - Marauder - 'Juggernaut', 'Berserker', 'Chieftain',
    원 - Witch - 'Necromancer', 'Elementalist', 'Occultist'
    원 - Ranger - 'Raider', 'Deadeye', 'Pathfinder'
    원 - Templar - 'Inquisitor', 'Hierophant', 'Guardian'
    사이온 - Scion - 'Ascendant'
```
방식 2: 세부 클래스(전직 후)에 따라 labeling (딜 위주, 탱킹 위주, 보조 위주 - 게임 공부를 더 하고 나누는 걸로. EDA 하다보면 아이디어를 얻을수도 있고,)

    'Slayer' : 양손 무기, 생존력 특화 Dualist
    'Gladiator' : 쌍수 혹은 한손-방패 Dualist
    'Champion' : 적 디버프, 자신 버프 Dualist
    'Assassin' : 출혈, 독 등 이용한 치명 딜 위주 Shadow
    'Saboteur' : 트랩, 마인 활용 Shadow
    'Trickster' : 기동성, 회피, 공격속도 위주 Shadow
    'Juggernaut' : 탱킹 중심 Marauder
    'Berserker' : 흡혈이나 피해 비례 데미지 등 광전사 Marauder
    'Chieftain' : 화염, 적 디버프 이용 Marauder
    'Necromancer' : 소환술 특화 Witch
    'Elementalist' : 원소 피해 특화 Witch
    'Occultist' : 카오스 피해, 저주 특화 Witch
    'Raider' : 회피, 처치 후 얻는 버프를 이용하는 Ranger
    'Deadeye' : 치명타, 딜 중심 Ranger
    'Pathfinder' : 플라스크(물약), 치유효과 강화, 디버프 제거, 원소, 카오스 이용 Ranger
    'Inquisitor' : 원소 치명타, 마나 재생, 자신 버프 Templer
    'Hierophant' : 토템 및 마나 변환, 받은 피해 마나로 감소 등의 Templer
    'Guardian' : 서포트 및 보호(방어버프, 디버프제거, 치유 등) 특화 Templer
    'Ascendant' : 타 클래스들의 조합하는 Scion. 성능은 다소 떨어질 수 있음.

방식 3: 특정 스탯 한 가지 / 혹은 두 가지 / 세 가지

    'Marauder', 'Witch', 'Ranger'
    'Duelist', 'Shadow', 'Templar', 'Scion'

In [20]:
df_byladder['class'].unique()

array(['Necromancer', 'Deadeye', 'Raider', 'Occultist', 'Pathfinder',
       'Guardian', 'Champion', 'Gladiator', 'Inquisitor', 'Elementalist',
       'Slayer', 'Berserker', 'Juggernaut', 'Ascendant', 'Trickster',
       'Hierophant', 'Chieftain', 'Saboteur', 'Assassin', 'Ranger',
       'Marauder', 'Scion', 'Witch', 'Duelist', 'Shadow', 'Templar'],
      dtype=object)

전직 전과 전직 후 모두 섞여있는 것 또한 볼 수 있다. 이에 따라 이를 기준으로 나눌 수도 있다.

전직 전과 후로 찍을 수 있는 스킬이 다를 것이기 때문이다. (물론 젬이 필요하다고 했기 때문에 그것이 없어서 그 직업을 운용중일 지도 모른다)

# class labeling

## 전직 전과 후로 나누기

In [22]:
df

rank   dead  online                    name  level        class  \
0          1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1          1  False   False        RaizNeverFirstQT    100  Necromancer   
2          1  False   False  GucciStreamerAdvantage    100  Necromancer   
3          1  False   False             ChiroxPrime    100       Slayer   
4          2  False   False     Cool_NecroIsFineNow    100      Deadeye   
...      ...    ...     ...                     ...    ...          ...   
59771  14999   True   False       ПроклятьеРекласта     89  Necromancer   
59772  15000  False   False       IshibashiSummoner     94  Necromancer   
59773  15000  False   False                 BLively     73       Slayer   
59774  15000  False   False                vawddvaw     89    Gladiator   
59775  15000  False   False                 Reselin     53  Necromancer   

                                                      id  experience  \
0      3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1      8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2      c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3      c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4      24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
...                                                  ...         ...   
59771  d33b4f6e08c10e365765f9a36a8f36d561fd1d86f10ee1...  1786695653   
59772  5764cfa387e0a87a4bebc1a3c5017e92de8bbb06445c47...  2720770853   
59773  9ac75ab75a47cee8a9dfb0a31912df8909720a8b20f9f1...   495034029   
59774  cf02dfc0c90b2df9c7ac76bbedd91e93c2a8a2ca629c15...  1786638972   
59775  f7ffda5ca2490546344d32930693f8299930ce9451df4a...    77909816   

            account  challenges        twitch              ladder  
0            TheTzn          40        thetzn           Harbinger  
1            RaizQT          22        raizqt    SSF Harbinger HC  
2       GucciPradas          35   guccipradas  Hardcore Harbinger  
3          Chiroxun          26           NaN       SSF Harbinger  
4          cooltail          22           NaN           Harbinger  
...             ...         ...           ...                 ...  
59771  Romanchik002          18  romanchik002  Hardcore Harbinger  
59772  Ishibashisan          33           NaN           Harbinger  
59773        ruby01          19           NaN       SSF Harbinger  
59774        Lecarn          24        lecarn  Hardcore Harbinger  
59775      woworpoe          19           NaN    SSF Harbinger HC  

[59776 rows x 12 columns]

In [23]:
df['전직여부'] = 0

In [25]:
pre_classes = ['Duelist', 'Shadow', 'Marauder', 'Witch', 'Ranger', 'Templar', 'Scion']

In [29]:
# 전직 여부에 따라 세부 클래스로 전직 전이면 0, 전직 후면 1

if df.loc[df['class'], :].isin(pre_classes):
    df['전직여부'] = 0
else:
    df['전직여부'] = 1
    
    
# def pre_class(x) :
#   if x in pre_class:
#     return str(x).split("(")[1].replace(")","") 
#   return str(x)

KeyError: "None of [Index(['Necromancer', 'Necromancer', 'Necromancer', 'Slayer', 'Deadeye',\n       'Deadeye', 'Gladiator', 'Inquisitor', 'Necromancer', 'Raider',\n       ...\n       'Raider', 'Gladiator', 'Necromancer', 'Saboteur', 'Berserker',\n       'Necromancer', 'Necromancer', 'Slayer', 'Gladiator', 'Necromancer'],\n      dtype='object', length=59776)] are in the [index]"

In [33]:
df['전직여부'] = np.where(df['class'].isin(pre_classes), 0, 1)

In [35]:
df.head(10)

rank   dead  online                    name  level        class  \
0     1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1     1  False   False        RaizNeverFirstQT    100  Necromancer   
2     1  False   False  GucciStreamerAdvantage    100  Necromancer   
3     1  False   False             ChiroxPrime    100       Slayer   
4     2  False   False     Cool_NecroIsFineNow    100      Deadeye   
5     2  False   False  SteelStreamerAdvantage    100      Deadeye   
6     2  False   False   Grom_WhyGodThreeCmWhy    100    Gladiator   
7     2  False   False          beobio_capitan    100   Inquisitor   
8     3  False   False    Rami_The_Negromaniac    100  Necromancer   
9     3  False   False           FutonBlewAway    100       Raider   

                                                  id  experience  \
0  3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1  8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2  c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3  c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4  24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
5  51acb8e6fac3cfaa7ff848151dad01b0a6f350e7525543...  4250334444   
6  5d8b80b2bdd30f3dba1036a6791187562758935e8d4723...  4250334444   
7  65c58f9b186e1fbe9f927484f3f178cdb9ee0dc2427289...  4250334444   
8  00c0030fc548cc4c0d804564f8fb356610044b7e6d484f...  4250334444   
9  46e56572ff6e403f45fac597240e13e9af225f06ca4e52...  4250334444   

         account  challenges        twitch              ladder  전직여부  
0         TheTzn          40        thetzn           Harbinger     1  
1         RaizQT          22        raizqt    SSF Harbinger HC     1  
2    GucciPradas          35   guccipradas  Hardcore Harbinger     1  
3       Chiroxun          26           NaN       SSF Harbinger     1  
4       cooltail          22           NaN           Harbinger     1  
5      Steelmage          37  steelmage202  Hardcore Harbinger     1  
6     rocklegend          22       gromick    SSF Harbinger HC     1  
7  TishaVosemsot          40        beobio       SSF Harbinger     1  
8       rami1337          40      rami1337           Harbinger     1  
9        pleaGRC          40           NaN       SSF Harbinger     1

### 몇명이나 전직하지 않고 랭킹에 올랐을까?

In [41]:
df['전직여부'].value_counts()

1    59280
0      496
Name: 전직여부, dtype: int64

In [43]:
freq = df.groupby(['전직여부']).size() 
print(freq)

전직여부
0      496
1    59280
dtype: int64


### 그럼 전직하지 않은 사람들의 직업은?

In [44]:
df[df['전직여부'] == 0]['class'].value_counts()

Witch       119
Marauder     95
Templar      71
Duelist      65
Scion        52
Ranger       48
Shadow       46
Name: class, dtype: int64

위 결과를 보니 더더욱! 라벨링을 하는게 좋겠다는 생각이 드는데, 게임을 어떤 관점에서 볼지에 따라 나누는 방식을 달리할 수 있을 것이다.

포지션 자체인지 아니면 전체적인 스킬의 유형(사거리 등)에 관한 것인지 사람마다 중요시 하는 부분이 다르기 때문에... 

살펴보면 살펴볼 수록 이 게임은 사거리 자체보다는 포지션이나, 솔플 난이도 여부에 따라 갈리는게 나을 것 같다는 생각이 든다. 그래서 SSF 모드가 따로 있는 것이기도 하겠지.

### 전직하지 않은 사람들의 플레이모드는?

In [45]:
df[df['전직여부'] == 0]['ladder'].value_counts()

SSF Harbinger HC      412
SSF Harbinger          78
Hardcore Harbinger      6
Name: ladder, dtype: int64

독특하게도 일반 Harbinger 에는 없었고, 대부분 솔로 플레이모드에 많았다.

아무래도 전직이라는 건 세부 직업을 고르는 것, 즉 특화를 하는 것이기 때문에 다른 면에서는 부족해질 수 있다고 생각해보면 솔로플레이에서는 전직하지 않은 직업으로 플레이 하는게 더 나을 수도 있다는 생각이다.

물론 스킬을 찍는 포인트가 한정되어있다는 사실 안에 말이다.



###  전직하지 않은 사람들의 플레이모드 별 다른 컬럼을 비교해보자

In [47]:
df_pre_cl = df[df['전직여부'] == 0]
df_pre_cl

rank   dead  online             name  level     class  \
4483    1123   True   False          Venkula     88     Witch   
5230    1310  False   False  FrozenTheocracy     88  Marauder   
9003    2256  False   False       Wintersorg     85  Marauder   
9220    2311  False   False         Czardone     94    Ranger   
9571    2399  False   False      DuDuCia_SSF     90  Marauder   
...      ...    ...     ...              ...    ...       ...   
59635  14965  False   False          Bratakk     53   Templar   
59672  14975   True   False           zyxzcb     53     Scion   
59717  14986   True   False        flackoraz     53  Marauder   
59731  14989   True   False          Aralena     53     Scion   
59746  14993  False   False        CrankyMin     53     Witch   

                                                      id  experience  \
4483   ff2158b95cc2f345f58a749dc5b87479f3c6c869c2c56d...  1742518167   
5230   d8bee7876d1d73ee030c7d2574d0d2270c9446a5a2c786...  1661086034   
9003   7b93ae4d6bba754e6cebc50518b35f9cea736baf0b58e0...  1308994834   
9220   5cd690e7d5b2092c1c763782a9d20440edd4fd6bcac1c5...  2678680422   
9571   dd90b4d7630137bcc29e761c8d5da905054894071461ea...  1934009687   
...                                                  ...         ...   
59635  ab43cb5747cbcfdaeca5e92c4fd9379a8a4035cf564980...    78452021   
59672  130a9fc1363dbc9867d34b7e557e586400c0c8026f6681...    78307968   
59717  8a6eb406ad742208f1f51530be2cb76a5a6b6cf0894f3f...    78083773   
59731  dc4404dae50544f86092ad1755d0b6b0f4e8c5a8a567f7...    78059267   
59746  69e3b0c17b0d75089bbfd52052acb2aed4bb46b630c03d...    77984723   

             account  challenges         twitch              ladder  전직여부  
4483          Baacax          25            NaN    SSF Harbinger HC     0  
5230      Warlock279          24            NaN    SSF Harbinger HC     0  
9003      Sandman421          14            NaN    SSF Harbinger HC     0  
9220   azzmasterking          15            NaN  Hardcore Harbinger     0  
9571     shanghaibf4          21            NaN       SSF Harbinger     0  
...              ...         ...            ...                 ...   ...  
59635     WizardOfUz           8            NaN    SSF Harbinger HC     0  
59672   BlindMessiah          17            NaN    SSF Harbinger HC     0  
59717        Virchan          16            NaN    SSF Harbinger HC     0  
59731        Imladir          18  imladir_talen    SSF Harbinger HC     0  
59746     crankyzola           8            NaN    SSF Harbinger HC     0  

[496 rows x 13 columns]

In [48]:
df_pre_cl.describe()

rank       level    experience  challenges   전직여부
count    496.000000  496.000000  4.960000e+02  496.000000  496.0
mean   11900.975806   63.195565  3.064528e+08   15.183468    0.0
std     2798.356468    9.487475  3.707211e+08    6.326839    0.0
min     1123.000000   53.000000  7.798472e+07    5.000000    0.0
25%    10924.750000   55.000000  9.925252e+07   10.000000    0.0
50%    12580.500000   60.000000  1.559193e+08   14.000000    0.0
75%    14080.250000   70.000000  3.837313e+08   18.000000    0.0
max    14993.000000   94.000000  2.678680e+09   37.000000    0.0

평균 랭크 자체는 높은 편이아니고, 레벨의 최대 값은 94였다. 최대 레벨이 100인 점을 감안할 때 물음표가 생길 수 밖에 없는 부분이다. 게임을 정확히 이해하지 않고서는 풀기 어려운 의문인 것 같다.

반대로 전직을 한 사람들을 보자면,

In [49]:
df[df['전직여부'] == 1].describe()

rank         level    experience    challenges     전직여부
count  59280.000000  59280.000000  5.928000e+04  59280.000000  59280.0
mean    7458.507355     85.608553  1.838498e+09     25.848060      1.0
std     4321.827856     11.116002  1.089898e+09      8.400402      0.0
min        1.000000     53.000000  7.790982e+07      4.000000      1.0
25%     3716.000000     79.000000  7.992337e+08     20.000000      1.0
50%     7440.000000     90.000000  1.934066e+09     25.000000      1.0
75%    11190.000000     95.000000  2.876117e+09     33.000000      1.0
max    15000.000000    100.000000  4.250334e+09     40.000000      1.0

평균 랭크가 약 3천 이상 더 높고, 레벨도 평균적으로 더 높다. 하지만 경험치는 그들보다 평균적으로 더 적었고, 시도도 더 많이 한 것으로 나타났다.

이를 토대로 합리적의심을 해보자면,